infer/internal/ende/ende.go

```diff
@@ -338,7 +361,38 @@ const (
 	isEmptyArr = iota
 )
 
-// Mark a encoder as generating values only.
+func flattenAssets(a any) (resource.PropertyValue, bool) {
+	if aMap, ok := a.(map[string]any); ok {
+		rawAsset, hasAsset := aMap[types.AssetSignature]
+		rawArchive, hasArchive := aMap[types.ArchiveSignature]
+
+		if hasAsset && hasArchive {
+			panic(`Encountered both an asset and an archive in the same AssetOrArchive. This
+should never happen. Please file an issue at https://github.com/pulumi/pulumi-go-provider/issues.`)
```

Comment:
I believe that this is when we are serializing. We need to serialize `AssetOrArchive` into either an asset or an archive, and we can't do that. We can't compress `struct { Asset, Archive }` into an `Asset` or `Archive` without losing information. It is *always* a provider bug and right now `panic`ing is a reasonable option. It would be better to return an error, but it's not clear how.

Reply:
Platform will not create a new type that has our sigs (we assume that sigs won't conflict by accident).

Reply:
It was the other way round - when we're parsing - we're adding a panic if we recieve something which contains both asset or archive information. My suggestion here would be to just check if it's an asset first, then check if it's an archive. The code below would already do that if the panic removed, if I read it correctly, though it prefers archive over asset.

Reply:
We should panic on ambiguity, as it is always a provider or framework bug. The user will never see the ambiguity and the engine can only pass ambiguous information if it copies the sigs we use, which it should not.

Reply:
I agree if it's user or framework error, but this data is coming from the engine – for which we don't have a super-good specification of what they might send, so I'd consider being defensive over making contract assertions. But perhaps it's reasonable here to fail hard because it would be a weird change engine side that _hopefully_ they wouldn't make because it's likely not backward compatible (not that we define backward compatibility in our protocols either 😞).

Reply:
There is **no** remotely backwards compatible change (beyond copying the sigs we created so they wouldn't be copied, on purpose or by accident) that would allow this to be an engine bug. It will **always** be a provider author or framework bug, and it should panic to allow the responsible party to fix it, instead of doing the wrong thing for end users.